lib/sparse/src/index/posting_list.rs

```diff
@@ -24,6 +41,70 @@ impl PostingList {
         }
         posting_list.build()
     }
+
+    /// Upsert a posting element into the posting list.
+    ///
+    /// Worst case is adding a new element at the end of the list with a very large weight.
+    /// This forces to propagate it as potential max_next_weight to all the previous elements.
+    pub fn upsert(&mut self, mut posting_element: PostingElement) {
+        // find insertion point in sorted posting list
+        let index = self
+            .elements
+            .binary_search_by_key(&posting_element.record_id, |e| e.record_id);
+
+        let modified_index = match index {
+            Ok(found_index) => {
+                // Update existing element for the same id
+                let element = &mut self.elements[found_index];
+                if element.weight == posting_element.weight {
+                    // no need to update anything
+                    return;
+                }
+                element.weight = posting_element.weight;
+                found_index
+            }
+            Err(insert_index) => {
+                // Compute `max_next_weight` based on the element currently at `insert_index`
+                let next_element_max_weight = self
```

Comment:
Would we skip all this calculation and applying of `next_element_max_weight` if we called `propagate_max_next_weight_to_the_left(modified_index + 1)`?